swift/apple/FirezoneNetworkExtension/Adapter.swift

```diff
@@ -83,227 +70,145 @@ class Adapter {
-      guard let self = self else { return }
-
-      self.logger.log("Adapter.start")
-      guard case .stoppedTunnel = self.state else {
-        logger.error("\(#function): Invalid Adapter state")
-        completionHandler(.invalidState)
-        return
-      }
+    logger.log("Adapter.start")
+    guard case .tunnelStopped = self.state else {
+      logger.error("\(#function): Invalid Adapter state")
+      completionHandler(.invalidState)
+      return
+    }
 
-      self.callbackHandler.delegate = self
+    callbackHandler.delegate = self
 
-      if self.connlibLogFolderPath.isEmpty {
-        self.logger.error("Cannot get shared log folder for connlib")
-      }
+    if connlibLogFolderPath.isEmpty {
+      logger.error("Cannot get shared log folder for connlib")
+    }
 
-      self.logger.log("Adapter.start: Starting connlib")
-      do {
-        // We can only get the system's default resolvers before connlib starts, and then they'll
-        // be overwritten by the ones from connlib. So cache them here for getSystemDefaultResolvers
-        // to retrieve them later.
-        self.callbackHandler.setSystemDefaultResolvers(
-          resolvers: Resolv().getservers().map(Resolv.getnameinfo)
-        )
-        self.state = .startingTunnel(
-          session: try WrappedSession.connect(
-            self.controlPlaneURLString,
-            self.token,
-            self.getOrCreateFirezoneId(from: self.firezoneIdFileURL),
-            self.getDeviceName(),
-            self.getOSVersion(),
-            self.connlibLogFolderPath,
-            self.logFilter,
-            self.callbackHandler
-          ),
-          onStarted: completionHandler
+    self.logger.log("Adapter.start: Starting connlib")
+    do {
+      // Grab a session pointer
+      let session =
+        try WrappedSession.connect(
+          apiURL,
+          token,
+          DeviceMetadata.getOrCreateFirezoneId(logger: self.logger),
+          DeviceMetadata.getDeviceName(),
+          DeviceMetadata.getOSVersion(),
+          connlibLogFolderPath,
+          logFilter,
+          callbackHandler
         )
-      } catch let error {
-        self.logger.error("Adapter.start: Error: \(error)")
-        packetTunnelProvider?.handleTunnelShutdown(
-          dueTo: .connlibConnectFailure,
-          errorMessage: error.localizedDescription)
-        self.state = .stoppedTunnel
-        completionHandler(AdapterError.connlibConnectError(error))
+      // Attempt to set DNS right away
+      if let jsonResolvers = try? String(
+        decoding: JSONEncoder().encode(getSystemDefaultResolvers()), as: UTF8.self
+      ).intoRustString() {
+        session.setDns(jsonResolvers)
       }
+      // Update our internal state
+      self.state = .tunnelStarted(session: session)
+
+      // Start listening for network change events. The first few will be our
+      // tunnel interface coming up, but that's ok -- it will trigger a `set_dns`
+      // connlib.
```

Comment:
```suggestion
      // tunnel interface coming up, but that's ok -- connlib will behave correctly anyway
```

Also, does this mean there's a gap where network events could be accidentally dropped during startup? Does that matter?